File: src/scene/fixtures/mail.js
import * as THREE from "three";
import { toon, tagSprite } from "../kit.js";

/**
 * the mailroom
 *
 * Intake is the mouth of the pipeline and the only stage the room could not see:
 * things get captured, cached, and never filed, and nothing anywhere said so.
 * This is a pigeonhole per feed, stacked with what is sitting in it.
 *
 * The one visual decision that matters: WHEN THE SUMMARY IS STALE, THE ROOM
 * SHOWS THE STALENESS WHERE THE COUNT WOULD HAVE GONE. A stale count rendered as
 * a current one is a lie you cannot see, which is worse than no mailroom at all.
 * So a stale cabinet loses its numbers entirely, its paper goes amber, and a red
 * band runs across it saying why.
 *
 * Three appearances that must never converge:
 *   a number    the last run covered this feed and this much is waiting
 *   "clear"     genuinely nothing waiting, in green, and it means it
 *   "unknown"   nobody could count it (email is a live mailbox) or the source
 *               itself is broken. Hatched dark, never an empty shelf
 */

export const id = "mail";
export const title = "the mailroom";
export const wall = true;

const INK = "#4a3b33";
const CREAM = "#fdf6e8";
const ALERT = "#d1495b";
const AMBER = "#c07c2c";
const GO = "#3f9e6a";

const WOOD = "#e0b183";
const VOID = "#7d6047";

/** Above this the stack stops growing and the tag carries the real number. A
 *  pigeonhole with 50 notes in it should look worse than one with 4 without
 *  becoming a chimney that leaves the room. */
const MAX_SHEETS = 9;

const clip = (s, n) => (s.length > n ? s.slice(0, n - 1) + "…" : s);

/** Paper. Its own material per cabinet, never the shared cache, because stale
 *  paper is translucent and mutating a cached material would tint the office. */
function paperMaterial(stale) {
  return new THREE.MeshToonMaterial({
    color: stale ? AMBER : CREAM,
    transparent: stale,
    opacity: stale ? 0.55 : 1,
  });
}

export function build(ctx) {
  const slot = ctx.room?.wall;
  if (!slot) return null;

  const section = ctx.section || null;
  const ok = section?.state === "ok";
  // A source that is unconfigured, timed out or broken still gets a cabinet.
  // A mailroom that vanishes when intake is unreachable teaches you that no
  // mailroom means no mail, which is the exact thing this fixture is here for.
  const stale = ok ? !!section.stale : true;
  const banner = ok
    ? (stale ? clip(section.stale_reason || "the summary is not current", 62) : "")
    : `${section?.state || "no data"}: ${clip(section?.detail || "intake was never read", 46)}`;

  const holes = ok && section.pigeonholes?.length
    ? section.pigeonholes
    : [
        { key: "granola", label: "granola", waiting: null },
        { key: "capture", label: "mobile capture", waiting: null },
        { key: "email", label: "email", waiting: null },
      ];

  const W = slot.w;
  const H = slot.h;
  const BASE = 0.25;

  const group = new THREE.Group();
  group.position.set(slot.x, BASE, slot.z);

  const carcass = new THREE.Mesh(new THREE.BoxGeometry(W, H, 0.22), toon(WOOD));
  carcass.position.set(0, H / 2, 0.11);
  carcass.userData.fixture = { id, payload: { focus: null } };
  group.add(carcass);

  // The headline, hung above the cabinet at pod-sign scale so it is the thing
  // you read from the far end of the room rather than something you must walk to.
  const head = tagSprite(stale ? "MAILROOM · STALE" : "the mailroom", {
    bg: stale ? ALERT : INK,
    fg: CREAM,
    scale: 1.35,
  });
  head.position.set(0, H + 0.62, 0.3);
  group.add(head);

  // The band. Red, full width, and it sits ON the cabinet face where a person
  // looking for a number would look first.
  if (banner) {
    const band = new THREE.Mesh(
      new THREE.BoxGeometry(W - 0.16, 0.34, 0.06),
      new THREE.MeshBasicMaterial({ color: ALERT })
    );
    band.position.set(0, H - 0.26, 0.25);
    group.add(band);
    const why = tagSprite(banner, { bg: ALERT, fg: "#fff", scale: 0.82 });
    why.position.set(0, H - 0.26, 0.34);
    group.add(why);
  }

  const PAD = 0.12;
  const top = H - (banner ? 0.5 : 0.16);
  const cellW = (W - PAD * (holes.length + 1)) / holes.length;
  const cellH = top - 0.46;
  const paper = paperMaterial(stale);

  holes.forEach((hole, i) => {
    const cx = -W / 2 + PAD + cellW / 2 + i * (cellW + PAD);
    const floor = 0.46;

    const shelf = new THREE.Mesh(new THREE.BoxGeometry(cellW, cellH, 0.02), toon(VOID));
    shelf.position.set(cx, floor + cellH / 2, 0.23);
    group.add(shelf);

    const unknown = hole.waiting == null;
    const waiting = unknown ? 0 : hole.waiting;

    if (unknown) {
      // Hatched, not empty. Diagonal slats say "we could not look", where a bare
      // shelf would say "we looked and there was nothing".
      for (let s = 0; s < 5; s++) {
        const slat = new THREE.Mesh(
          new THREE.BoxGeometry(cellW * 0.66, 0.045, 0.02),
          new THREE.MeshBasicMaterial({ color: "#5c4632" })
        );
        slat.position.set(cx, floor + cellH * (0.2 + s * 0.14), 0.245);
        slat.rotation.z = -0.5;
        group.add(slat);
      }
    } else {
      const sheets = Math.min(waiting, MAX_SHEETS);
      for (let s = 0; s < sheets; s++) {
        const sheet = new THREE.Mesh(new THREE.BoxGeometry(cellW * 0.76, 0.05, 0.3), paper);
        // A hand-stacked pile, not a printer tray. Tiny jitter reads as paper.
        sheet.position.set(cx + (s % 2 ? 0.02 : -0.02), floor + 0.05 + s * 0.075, 0.3);
        sheet.rotation.y = (s % 3) * 0.02 - 0.02;
        group.add(sheet);
      }
    }

    // Where the count goes. Stale takes this slot; that is the whole point.
    const value = stale ? "stale" : unknown ? "unknown" : waiting ? String(waiting) : "clear";
    const bg = stale || (!ok && unknown) ? ALERT : unknown ? INK : waiting ? AMBER : GO;
    const tag = tagSprite(value, { bg, fg: "#fff", scale: 0.9 });
    tag.position.set(cx, floor + cellH - 0.24, 0.42);
    group.add(tag);

    const name = tagSprite(hole.label || hole.key, { bg: CREAM, fg: INK, scale: 0.78 });
    name.position.set(cx, 0.24, 0.42);
    group.add(name);

    // The click target is the whole pigeonhole, deep enough to catch a ray from
    // the room camera without needing a steady hand at desk scale.
    const pad = new THREE.Mesh(
      new THREE.BoxGeometry(cellW, cellH + 0.4, 0.6),
      new THREE.MeshBasicMaterial({ visible: false })
    );
    pad.position.set(cx, floor + cellH / 2 - 0.1, 0.5);
    pad.userData.fixture = { id, payload: { focus: hole.key } };
    group.add(pad);
  });

  return group;
}

/* ------------------------------------------------------------------ panel -- */

function line(api, label, value, cls) {
  const p = api.el("p", "log");
  p.append(api.el("b", null, label), " ");
  p.append(api.el("span", cls || null, String(value)));
  return p;
}

export function panel(payload, world, api) {
  const s = world?.sections?.mail;
  const wrap = api.el("div");

  if (!s || s.state !== "ok") {
    const why = {
      unconfigured: "No vault root is configured for this office, so intake was never asked.",
      missing: "There is no intake service at the configured vault root.",
      timeout: "Intake did not answer in time. This is NOT an empty mailroom: nobody could say what is in it.",
      unreadable: "Intake answered with something that was not its summary.",
      error: "Intake failed when asked for its summary.",
      unbuilt: "This source has not been built yet.",
    }[s?.state] || "There is no intake data in this snapshot.";
    const box = api.el("div", "issue hot");
    box.append(api.el("div", "issue-title", why));
    if (s?.detail) box.append(api.el("p", "p-detail", s.detail));
    wrap.append(box);
    return wrap;
  }

  // Staleness first, always, before a single count. Anything underneath it
  // describes a moment that has already passed.
  if (s.stale) {
    const box = api.el("div", "issue hot");
    box.append(api.el("div", "issue-title", "This summary is not current."));
    box.append(api.el("p", "p-detail", s.stale_reason || "intake did not say why"));
    box.append(api.el("p", "p-detail",
      "Every number below is what the LAST run decided, not what is true now. " +
      "Run a dry run to make them current."));
    wrap.append(box);
  } else {
    wrap.append(line(api, "current.", "the last run covers everything on disk", "ok"));
  }

  wrap.append(line(api, "last run", s.last_run || "never", s.last_run ? null : "bad"));
  if (s.dry_run != null) {
    wrap.append(line(api, s.dry_run ? "dry run" : "filing run",
      s.dry_run ? "decided everything, filed nothing" : "this run created issues"));
  }
  if (s.watermark) wrap.append(line(api, "watermark", s.watermark));

  wrap.append(api.el("h3", null, "pigeonholes"));
  for (const h of s.pigeonholes || []) {
    const unknown = h.waiting == null;
    const hot = s.stale || unknown || h.waiting > 0;
    const card = api.el("div", `issue${hot ? " hot" : ""}`);
    if (payload?.focus === h.key) card.style.borderColor = "rgba(74,59,51,0.55)";

    const top = api.el("div", "issue-top");
    top.append(api.el("span", "issue-num",
      s.stale ? "stale" : unknown ? "unknown" : String(h.waiting)));
    top.append(api.el("span", "issue-title", h.label || h.key));
    card.append(top);

    card.append(api.el("p", "p-detail", h.why || ""));
    const tags = api.el("div", "tags");
    if (h.on_disk != null) tags.append(api.el("span", "tag", `${h.on_disk} on disk`));
    tags.append(api.el("span", "tag", `${h.covered ?? 0} covered by the last run`));
    tags.append(api.el("span", "tag", `${h.filed ?? 0} filed as issues`));
    tags.append(api.el("span", `tag${h.in_last_run ? "" : " hot"}`,
      h.in_last_run ? "in the last run" : "not in the last run"));
    card.append(tags);
    wrap.append(card);
  }

  const c = s.counts || {};
  wrap.append(api.el("h3", null, "what the last run decided"));
  wrap.append(line(api, `${c.items ?? 0}`, "items decided"));
  wrap.append(line(api, `${c.would_file ?? 0}`, "a dry run says would be filed next"));
  wrap.append(line(api, `${c.filed ?? 0}`, "actually became issues"));
  wrap.append(line(api, `${c.cached ?? 0}`, "extractions cached on disk"));

  // Three different things. A single "not filed" number would hide the only
  // distinction that matters: one is a decision, one is a gate, one is a
  // FAILURE that looks exactly like an absence.
  const held = s.held || {};
  wrap.append(api.el("h3", null, "not filed, and not the same thing"));
  wrap.append(line(api, `${held.declined ?? 0}`, "declined: looked at, judged not actionable"));
  wrap.append(line(api, `${held.rate_limited ?? 0}`,
    "rate limited: Granola answered 429 and the retries ran out. NOT an absence",
    held.rate_limited ? "bad" : null));
  wrap.append(line(api, `${held.no_transcript ?? 0}`,
    "no transcript on disk yet", held.no_transcript ? "wait" : null));
  for (const [reason, n] of Object.entries(held.blocked || {})) {
    wrap.append(line(api, `${n}`, `blocked: ${reason}`, n ? "bad" : null));
  }
  if (!Object.keys(held.blocked || {}).length) {
    wrap.append(line(api, "0", "blocked on anything else", "ok"));
  }

  // Deliberate exclusions are data. Named, so a skipped item can never be
  // mistaken for one that quietly fell out of the pipeline.
  const ex = s.excluded || [];
  wrap.append(api.el("h3", null, `deliberately excluded (${ex.length})`));
  if (!ex.length) {
    wrap.append(api.el("p", "empty", "Nothing is excluded. Everything on disk is eligible."));
  } else {
    wrap.append(api.el("p", "p-detail",
      "Chosen by hand, not by a classifier. These were seen and kept out on purpose."));
    for (const item of ex) {
      const card = api.el("div", "issue");
      card.append(api.el("div", "issue-title", item.title || item.id || "(unnamed)"));
      if (item.id) card.append(api.el("div", "issue-repo", item.id));
      wrap.append(card);
    }
  }
  return wrap;
}

/**
 * The ugly cases on purpose: a stale summary, a feed nobody can count, a feed
 * that was not in the last run, a live rate limit, and a real exclusion list.
 * `?demo=1` that only ever shows the happy path is a demo that hides the bugs.
 */
export function demo() {
  return {
    state: "ok",
    stale: true,
    stale_reason: "granola: 12 of 50 on disk; capture was not in the last run",
    last_run: "2026-08-25T18:30:44.256629+00:00",
    watermark: "2026-08-25",
    dry_run: true,
    pigeonholes: [
      {
        key: "granola", label: "granola", in_last_run: true, on_disk: 50,
        covered: 12, filed: 3, waiting: 38,
        why: "38 of 50 on disk are not in the last run's decisions",
      },
      {
        key: "capture", label: "mobile capture", in_last_run: false, on_disk: 2,
        covered: 0, filed: 0, waiting: 2,
        why: "this feed was not in the last run, so none of it has been looked at",
      },
      {
        key: "email", label: "email", in_last_run: true, on_disk: null,
        covered: 9, filed: 1, waiting: null,
        why: "email is a live mailbox, so nothing about it can be counted from disk. Unknown, not zero.",
      },
    ],
    counts: { items: 21, would_file: 6, filed: 3, cached: 72 },
    held: {
      declined: 11,
      blocked: { blocked_on_identity: 2, not_opted_in: 1 },
      rate_limited: 4,
      no_transcript: 2,
    },
    excluded: [
      { id: "1dc01aab-ec47-4a2f-bef3-d8ccdc452a3d", title: "Career options and self-worth with mentor" },
      { id: "1e8a370a-3070-4d09-9307-f589021a5142", title: "Discuss Principles + Values" },
      { id: "", title: "any title matching /standup/" },
    ],
  };
}

File: client/sources/mail.py
"""what intake has caught and not yet filed.

Intake is the mouth of the pipeline: Granola notes, mobile captures and email go
in, GitHub issues come out. It is also the one stage with no representation in
the room, and the one where things go missing quietly, because "captured" and
"filed" are two different states and only the second one is visible anywhere
else.

WHICH INTERFACE THIS USES, AND WHY
----------------------------------
`intake.py --summary --json`, run as a subprocess with a hard timeout.

It advertises itself as "cache-only status: no model calls, no network, no
writes", and that was measured before this file was written rather than taken on
trust: 0.08s wall clock on the real vault, cold. That is well inside what a
snapshot push can afford, so the advertised interface is the one used. A plain
run of intake (no --summary) reaches the Granola API and can take minutes, so it
is never invoked here, and neither is --file, which creates real GitHub issues.

Two small files are read directly alongside it, because --summary drops facts
this room needs:

  cache/last-run.json   `rate_limited` and `no_transcript`. Intake's own
                        docstring insists rate_limited is NOT an absence, and
                        summarise() does not carry it through. A mailroom that
                        cannot tell "nothing came" from "Granola said 429" is
                        the false-green this project exists to kill.
  state.json            the exclusion list itself. --summary reports a COUNT
                        from the last run, which is 0 whenever the last run
                        happened to exclude nothing, even while three meetings
                        stand permanently excluded. The names are the data;
                        deliberately skipped must never look like dropped.

Both are cheap local reads and both fail soft: a missing or torn file costs the
detail, never the snapshot.

STALENESS DOMINATES
-------------------
`stale` and `stale_reason` come straight from intake and are the headline. Every
count below them is subordinate: if the summary does not cover what is on disk,
the numbers describe a moment that has passed, and the room must render the
staleness where the count would have gone.
"""

from __future__ import annotations

import json
import os
import pathlib
import subprocess
import sys

from sources import _card

KEY = "mail"
# The human name of the fixture, fixed. A card whose title moves is a
# card the eye has to find again every time it is drawn.
TITLE = "Mail"

SERVICE = "_meta/services/intake"

# Measured at 0.08s. Ten seconds is not a budget, it is a tripwire: if intake
# ever starts reaching the network from --summary, this fires instead of hanging
# the push, and a timeout reports as its OWN state rather than as an empty room.
TIMEOUT = 10

# The three feeds, in the order they hang on the wall. `on_disk` can count the
# first two; email lives on a server and can never be counted from disk, so it
# is unknown rather than zero, and the room draws those differently.
FEEDS = [
    ("granola", "granola", True),
    ("capture", "mobile capture", True),
    ("email", "email", False),
]


def _root() -> pathlib.Path | None:
    v = os.environ.get("OFFICE_RUNTIME_ROOT", "").strip()
    return pathlib.Path(v).expanduser() if v else None


def _json_file(path: pathlib.Path) -> dict:
    """A local JSON file, or {}. Never raises: these are extra detail, not the
    snapshot, and one torn cache file must not cost the whole mailroom."""
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError, ValueError):
        return {}


def _summary(script: pathlib.Path) -> dict:
    """Run intake's own status interface. Returns {"error": ...} rather than
    raising, so every failure mode gets a name in the room."""
    try:
        proc = subprocess.run(
            [sys.executable, str(script), "--summary", "--json"],
            capture_output=True,
            text=True,
            timeout=TIMEOUT,
            cwd=str(script.parent),
        )
    except subprocess.TimeoutExpired:
        return {"error": "timeout", "detail": f"intake --summary did not answer in {TIMEOUT}s"}
    except OSError as exc:
        return {"error": "error", "detail": f"could not run intake: {exc}"[:200]}

    if proc.returncode != 0:
        detail = (proc.stderr or proc.stdout or "").strip().splitlines()
        return {"error": "error",
                "detail": f"intake --summary exited {proc.returncode}: "
                          f"{detail[-1] if detail else 'no output'}"[:300]}
    try:
        return json.loads(proc.stdout)
    except json.JSONDecodeError:
        return {"error": "unreadable",
                "detail": "intake --summary printed something that was not JSON"}


def _excluded(state: dict) -> list:
    """The operator's exclusion list, as names rather than a number.

    Stored as "<id>  <title>" strings. Split on the first run of whitespace: the
    id is one token, everything after it is what a person would recognise.
    """
    out = []
    for raw in (state.get("exclude") or {}).get("meetings", []):
        text = str(raw).strip()
        if not text:
            continue
        bits = text.split(None, 1)
        out.append({"id": bits[0], "title": bits[1].strip() if len(bits) > 1 else ""})
    for pattern in (state.get("exclude") or {}).get("title_patterns", []):
        if str(pattern).strip():
            out.append({"id": "", "title": f"any title matching /{pattern}/"})
    return out


def _pigeonholes(summary: dict, snap: dict, state: dict) -> list:
    """One per feed: what is sitting in it, and why it has not moved.

    `waiting` is what exists on disk minus what the last run covered. Intake
    counts a deliberately excluded item as covered, on purpose, so an exclusion
    does not make the room look permanently behind.
    """
    on_disk = summary.get("on_disk") or {}
    covered = snap.get("covered") or {}
    ran = set(summary.get("sources") or [])

    filed_by_source = {}
    for record in (state.get("filed") or {}).values():
        src = str((record or {}).get("source") or "")
        filed_by_source[src] = filed_by_source.get(src, 0) + 1

    holes = []
    for key, label, countable in FEEDS:
        seen = int(covered.get(key, 0) or 0)
        hole = {
            "key": key,
            "label": label,
            "in_last_run": key in ran,
            "on_disk": int(on_disk[key]) if countable and key in on_disk else None,
            "covered": seen,
            "filed": filed_by_source.get(key, 0),
            "waiting": None,
            "why": "",
        }
        if hole["on_disk"] is None:
            hole["why"] = ("email is a live mailbox, so nothing about it can be counted "
                           "from disk. Unknown, not zero.")
        else:
            hole["waiting"] = max(hole["on_disk"] - seen, 0)
            if not hole["in_last_run"]:
                hole["why"] = "this feed was not in the last run, so none of it has been looked at"
            elif hole["waiting"]:
                hole["why"] = f"{hole['waiting']} of {hole['on_disk']} on disk are not in the last run's decisions"
            else:
                hole["why"] = "everything on disk was covered by the last run"
        holes.append(hole)
    return holes


def read() -> dict:
    root = _root()
    if root is None:
        return {"state": "unconfigured",
                "detail": "OFFICE_RUNTIME_ROOT is not set, so there is no vault to read intake from"}
    service = root / SERVICE
    script = service / "intake.py"
    if not script.exists():
        return {"state": "missing", "detail": f"no intake at {script}"}

    summary = _summary(script)
    if summary.get("error"):
        # A timeout is its own state and says so. It is emphatically not an
        # empty mailroom: it means nobody could answer the question.
        return {"state": summary["error"], "detail": summary.get("detail", "")}

    snap = _json_file(service / "cache" / "last-run.json")
    state = _json_file(pathlib.Path(os.environ.get("INTAKE_STATE") or (service / "state.json")))

    stale = bool(summary.get("stale"))
    return {
        "state": "ok",
        # The headline. Everything below is subordinate to these two.
        "stale": stale,
        "stale_reason": str(summary.get("stale_reason") or "")[:400] if stale else "",
        "last_run": summary.get("last_run"),
        "watermark": summary.get("watermark"),
        # The last run's own honesty about itself: a --file run and a dry run
        # decide the same things and do very different amounts about it.
        "dry_run": snap.get("dry_run"),
        "pigeonholes": _pigeonholes(summary, snap, state),
        "counts": {
            "items": int(summary.get("items") or 0),
            "would_file": int(summary.get("would_file") or 0),
            "filed": int(summary.get("filed") or 0),
            "cached": int(summary.get("cached") or 0),
        },
        # Three different things, kept apart on purpose. Declined is a decision,
        # blocked is a gate, rate_limited is a failure that LOOKS like an
        # absence. Adding them up would erase the only distinction that matters.
        "held": {
            "declined": int(summary.get("declined") or 0),
            "blocked": {str(k): int(v) for k, v in (summary.get("blocked") or {}).items()},
            "rate_limited": int(snap.get("rate_limited") or 0),
            "no_transcript": int(snap.get("no_transcript") or 0),
        },
        # Deliberate, named, and never folded into the numbers above.
        "excluded": _excluded(state),
    }


TROUBLE = {
    "unconfigured": ("not configured", 0),
    "missing": ("intake is not installed", 1),
    "timeout": ("intake did not answer", 1),
    "unreadable": ("intake printed something that was not JSON", 1),
    "error": ("intake would not run", 1),
}

# Held back for three different reasons, and they are never added up. Declined
# is a decision, blocked is a gate, rate limited is a failure that LOOKS like an
# absence, and no transcript is a meeting nobody wrote down.
HELD = (("declined", "declined"), ("blocked", "blocked"),
        ("rate_limited", "rate limited"), ("no_transcript", "no transcript"))


def card(data: dict) -> dict:
    """What is sitting in the mailroom, and whether the count can be trusted.

    Staleness outranks every number below it. If the last run does not cover
    what is on disk then the counts describe a moment that has passed, so the
    headline says so where the count would have gone.
    """
    if data.get("state") != "ok":
        return _card.trouble(TITLE, data.get("state"), data.get("detail"), TROUBLE)

    holes = data.get("pigeonholes") or []
    waiting = sum(int(h.get("waiting") or 0) for h in holes)
    stale = bool(data.get("stale"))

    if stale:
        headline = "behind: " + (data.get("stale_reason")
                                 or "the last run does not cover what is on disk")
    elif waiting:
        headline = f"{_card.count(waiting)} waiting to be filed"
    else:
        headline = "nothing waiting to be filed"

    facts = []
    for hole in holes:
        n = hole.get("waiting")
        filed = _card.count(hole.get("filed") or 0)
        if n is None:
            # Email lives on a server. Unknown, never zero, and drawn as such.
            value, tone = f"unknown, {filed} filed", "dim"
        else:
            n = int(n)
            value = f"{_card.count(n)} waiting, {filed} filed"
            tone = "warn" if n else "ok"
        facts.append(_card.fact(hole.get("label") or hole.get("key") or "feed", value, tone))

    last_run = data.get("last_run")
    facts.append(_card.fact("last run", _card.ago(last_run) or "never",
                            "warn" if stale else "dim"))

    held = data.get("held") or {}
    counted = {**held, "blocked": sum(int(v or 0) for v in (held.get("blocked") or {}).values())}
    parts = [f"{_card.count(int(counted.get(key) or 0))} {label}"
             for key, label in HELD if int(counted.get(key) or 0)]
    if parts:
        facts.append(_card.fact("held back", ", ".join(parts),
                                "warn" if int(counted.get("rate_limited") or 0) else "dim"))

    return _card.build(TITLE, headline, waiting, _card.zulu(last_run), facts)
